providers/providers_manager.go

```diff
@@ -296,6 +319,27 @@ func (pm *ProviderManager) GetProviders(ctx context.Context, k []byte) ([]peer.A
 	}
 }
 
+// GetProvidersForPrefix returns the set of providers with the given prefix, as well
+// as the full key they provide.
+func (pm *ProviderManager) GetProvidersForPrefix(ctx context.Context, k []byte) (map[peer.ID][][]byte, error) {
```

Comment:
Why do we return `map[peer.ID][][]byte` here? Wouldn't it be better to return `map[[]byte]peer.ID`? In the end we only care about one provider record associated with the CID. If we have a `map[peer.ID][][]byte` we need to iterate on the full map.
How does it integrate with provider record decryption?

Reply:
the map value is what keys with the prefix the peer has. it's possible the peer has multiple keys with the same prefix, so that's why it's `[][]byte`. 

for provider record decryption, the `peer.ID` is actually encrypted.

Reply:
But when requesting a specific key, we expect `Pr[many peerID for one CID matching prefix] >> Pr[many CIDs matching prefix provided by the same peerID]` as if a provider provides multiples CIDs, they will be stored at *random* locations in the DHT, hence we don't expect to have multiple hits for the same content provider for a single request (targeting a specific location). 

I think that it would make more sense to have a map: second hash (dht identifier) -> list of encrypted peer.ID

What do you think?

Reply:
okay, I guess even though that case is unlikely, it still could potentially happen (especially if the prefix length chosen isn't that short). what would you suggest doing in the case where there are multiple matching records for the same peer? just pick one at random?